controller/konnect/konnectextension_controller_secrets.go

```diff
@@ -0,0 +1,79 @@
+package konnect
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	konnectv1alpha1 "github.com/kong/kubernetes-configuration/api/konnect/v1alpha1"
+)
+
+const (
+	// SecretKonnectDataPlaneCertificateLabel is the label to mark that the secret is used as a Konnect DP certificate.
+	// A secret must have the label to be watched by the KonnectExtension reconciler.
+	SecretKonnectDataPlaneCertificateLabel = "konghq.com/konnect-dp-cert" //nolint:gosec
+)
+
+func listKonnectExtensionsBySecret(ctx context.Context, cl client.Client, obj client.Object) ([]konnectv1alpha1.KonnectExtension, error) {
+	s, ok := obj.(*corev1.Secret)
+	if !ok {
+		return nil, fmt.Errorf("expect object to be secret, actually %T", obj)
+	}
+	l := &konnectv1alpha1.KonnectExtensionList{}
+	err := cl.List(
+		ctx, l,
+		client.InNamespace(s.Namespace),
+		client.MatchingFields{
+			IndexFieldKonnectExtensionOnSecrets: s.Name,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return l.Items, nil
+
+}
+
+func secretUsedByKonnectExtension(cl client.Client) func(obj client.Object) bool {
+	return func(obj client.Object) bool {
+		konnectExtensions, err := listKonnectExtensionsBySecret(context.Background(), cl, obj)
+		if err != nil {
+			return false
+		}
+		return lo.ContainsBy(konnectExtensions, func(ke konnectv1alpha1.KonnectExtension) bool {
+			return ke.Spec.DataPlaneClientAuth != nil &&
+				ke.Spec.DataPlaneClientAuth.CertificateSecret.CertificateSecretRef != nil &&
+				ke.Spec.DataPlaneClientAuth.CertificateSecret.CertificateSecretRef.Name == obj.GetName()
+		})
+	}
+}
+
+func konnectExtensionReconcileRequestsForSecret(cl client.Client) func(context.Context, client.Object) []reconcile.Request {
+	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+		konnectExtensions, err := listKonnectExtensionsBySecret(ctx, cl, obj)
```

Comment:
Can we type assert at the top of this enqueue func and use the typed `Secret` throughout this func?

```suggestion
	return func(ctx context.Context, obj client.Object) []reconcile.Request {
		s, ok := obj.(*corev1.Secret)
		if !ok {
			return nil, fmt.Errorf("expect object to be secret, actually %T", obj)
		}
		konnectExtensions, err := listKonnectExtensionsBySecret(ctx, cl, s)
```